java/feathub-udf/flink-udf/src/main/java/com/alibaba/feathub/flink/udf/aggregation/AggFuncWithLimit.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2022 The Feathub Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.feathub.flink.udf.aggregation;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.common.typeinfo.Types;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.Preconditions;
+
+import java.util.List;
+
+/**
+ * Aggregation function decorator that only aggregates up to `limit` number of most recent records.
+ */
+public class AggFuncWithLimit<IN_T, PRE_OUT_T, OUT_T, PRE_ACC_T, ACC_T>
+        implements AggFunc<List<Tuple2<IN_T, Long>>, OUT_T, RawDataAccumulator<IN_T>> {
+    private final PreAggFunc<IN_T, PRE_OUT_T, PRE_ACC_T> preAggFunc;
```

Comment:
Since `AggFuncWithLimit` stores all values in the RawDataAccumulator, it effectively disables the pre-aggregation optimization. Then it is hard to understand why this class needs `preAggFunc`.